### PR TITLE
[Merged by Bors] - Implement demo command

### LINKED ISSUE #56: POC Demo trino-superset-s3
We want to create a first demo with data. For a simple stack we decided to use Trino + S3 + Superset.

- [x] Create a stack
- [x] Evaluate making data available
       - create k8s job
- [x] Testing data 
       - create ConfigMap

### PR DESCRIPTION
**Pull request by sbernauer**
## Description

Proudly presenting the first Stackable demo of the world :partying_face: 
Works best with https://github.com/stackabletech/superset-operator/pull/247.
If you don't increasing the Superset timeout the charts may time out. In that case simply refresh them **one by one while there is no other query running on Trino (check via Trino webui)**.

image
image

To invoke run (from feature branch)
```
sbernauer@sbernauer:~/stackabletech/stackablectl$ cargo r -- --additional-demos-file demos/demos-v1.yaml demo install trino-taxi-data
    Finished dev [unoptimized + debuginfo] target(s) in 0.06s
     Running `target/debug/stackablectl --additional-demos-file demos/demos-v1.yaml demo install trino-taxi-data`
[INFO ] Installing demo trino-taxi-data
[WARN ] Failed to parse demo list from https://raw.githubusercontent.com/stackabletech/stackablectl/main/demos/demos-v1.yaml: missing field `demos` at line 1 column 4
[INFO ] Installing stack trino-superset-s3
[INFO ] Installing release 22.06
[INFO ] Installing airflow operator in version 0.4.0
[INFO ] The release airflow-operator in version 0.4.0 is already installed, not installing it
[INFO ] Installing commons operator in version 0.2.0
[INFO ] The release commons-operator in version 0.2.0 is already installed, not installing it
[INFO ] Installing druid operator in version 0.6.0
[INFO ] The release druid-operator in version 0.6.0 is already installed, not installing it
[INFO ] Installing hbase operator in version 0.3.0
[INFO ] The release hbase-operator in version 0.3.0 is already installed, not installing it
[INFO ] Installing hdfs operator in version 0.4.0
[INFO ] The release hdfs-operator in version 0.4.0 is already installed, not installing it
[INFO ] Installing hive operator in version 0.6.0
[INFO ] The release hive-operator in version 0.6.0 is already installed, not installing it
[INFO ] Installing kafka operator in version 0.6.0
[INFO ] The release kafka-operator in version 0.6.0 is already installed, not installing it
[INFO ] Installing nifi operator in version 0.6.0
[INFO ] The release nifi-operator in version 0.6.0 is already installed, not installing it
[INFO ] Installing opa operator in version 0.9.0
[INFO ] The release opa-operator in version 0.9.0 is already installed, not installing it
[INFO ] Installing secret operator in version 0.5.0
[INFO ] The release secret-operator in version 0.5.0 is already installed, not installing it
[INFO ] Installing spark-k8s operator in version 0.3.0
[INFO ] The release spark-k8s-operator in version 0.3.0 is already installed, not installing it
[INFO ] Installing superset operator in version 0.5.0
[INFO ] The release superset-operator in version 0.5.0 is already installed, not installing it
[INFO ] Installing trino operator in version 0.4.0
[INFO ] The release trino-operator in version 0.4.0 is already installed, not installing it
[INFO ] Installing zookeeper operator in version 0.10.0
[INFO ] The release zookeeper-operator in version 0.10.0 is already installed, not installing it
[INFO ] Installing components of stack trino-superset-s3
[INFO ] The release minio-trino in version 4.0.5 is already installed, not installing it
[INFO ] The release postgresql-hive in version 10.16.2 is already installed, not installing it
[INFO ] The release postgresql-superset in version 11.0.0 is already installed, not installing it
[INFO ] Installed stack trino-superset-s3
[INFO ] Installing components of demo trino-taxi-data
[INFO ] Installed demo trino-taxi-data. Use "stackablectl services list" to list the installed services




sbernauer@sbernauer:~/stackabletech/stackablectl$ stackablectl services list
 PRODUCT   NAME         NAMESPACE  ENDPOINTS                                     EXTRA INFOS                          
                                                                                                                      
 hive      hive         default    hive                172.18.0.4:32658                                               
                                   metrics             172.18.0.4:30745                                               
                                                                                                                      
 opa       opa          default    http                http://172.18.0.2:31324                                        
                                                                                                                      
 superset  superset     default    external-superset   http://172.18.0.2:32716   Admin user: admin, password: admin   
                                                                                                                      
 trino     trino        default    coordinator-http    http://172.18.0.5:32128                                        
                                   coordinator-metrics 172.18.0.5:31199                                               
                                   coordinator-https   https://172.18.0.5:32721                                       
                                                                                                                      
 minio     minio-trino  default    http                http://172.18.0.4:31026   Third party service                  
                                   console-http        http://172.18.0.4:30354   Admin user: root, password: rootroot
```

## Review Checklist
- [ ] Code contains useful comments
- [ ] (Integration-)Test cases added (or not applicable)
- [ ] Documentation added (or not applicable)
- [ ] Changelog updated (or not applicable)
- [ ] Cargo.toml only contains references to git tags (not specific commits or branches)

Once the review is done, comment `bors r+` (or `bors merge`) to merge. [Further information](https://bors.tech/documentation/getting-started/#reviewing-pull-requests)

### CODE REVIEW

File: demos/demos-v1.yaml [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+    labels:
+      - trino
+      - superset
+      - minio
+      - s3
+      - ny-taxi-data
+    manifests:
+      - plainYaml: https://raw.githubusercontent.com/stackabletech/stackablectl/demos/demos/trino-taxi-data/load-test-data.yaml
+      - plainYaml: https://raw.githubusercontent.com/stackabletech/stackablectl/demos/demos/trino-taxi-data/create-table-in-trino.yaml
+      - plainYaml: https://raw.githubusercontent.com/stackabletech/stackablectl/demos/demos/trino-taxi-data/setup-superset.yaml
```
Comment (sbernauer) [inline demos/demos-v1.yaml:15]: Change to main branch before merging

File: demos/trino-taxi-data/setup-superset.yaml [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+kind: Job
+metadata:
+  name: setup-superset
+spec:
+  template:
+    spec:
+      containers:
+        - name: setup-superset
+          image: python:3.10-slim
+          command: ["bash", "-c", " apt update && apt install -y curl && curl -o superset-assets.zip https://raw.githubusercontent.com/stackabletech/stackablectl/demos/demos/trino-taxi-data/superset-assets.zip && pip install requests==2.22.0 && python /tmp/script/script.py"]
```
Comment (sbernauer) [inline demos/trino-taxi-data/setup-superset.yaml:12]: Change to main branch before merging

**Review by maltesander — CHANGES_REQUESTED** [review]
Just one typo. Works great!
image
Just ran in the superset timeout, but nothing we can do about that here.
```
malte@mdesktop ~/d/w/stackablectl (demos) [2]> target/debug/stackablectl demo list
[WARN ] Failed to parse demo list from https://raw.githubusercontent.com/stackabletech/stackablectl/main/demos/demos-v1.yaml: missing field `demos` at line 1 column 4
DEMO                                STACKABLE STACK           DESCRIPTION
malte@mdesktop ~/d/w/stackablectl (demos)> target/debug/stackablectl demo describe trino-taxi-data
[WARN ] Failed to parse demo list from https://raw.githubusercontent.com/stackabletech/stackablectl/main/demos/demos-v1.yaml: missing field `demos` at line 1 column 4
[ERROR] Demo trino-taxi-data not found. Use `stackablectl demo list` to list the available demos.
malte@mdesktop ~/d/w/stackablectl (demos) [255]> 
```
The list and describe do not work for me?

File: docs/modules/ROOT/pages/commands/demo.adoc [hunk 3]
```diff
@@ -1,3 +1,129 @@
...
-Not implemented yet
+A demo is an end-to-end demonstration of the usage of the Stackable data platform.
+It is tied to a specific stack of the Stackable data platform, which will provide the required products for the demo.
+
+== Browse available demos
+To list the available demos, run the following command:
+
+[source,console]
+----
+$ stackablectl demos list
```
[the remark this answers is not in the document]
Comment (sbernauer) [reply docs/modules/ROOT/pages/commands/demo.adoc:11]: fixed, thx

**Comment by sbernauer** [comment]
As this PR is not merged you have to specify the additional demos file like
`cargo r -- --additional-demos-file demos/demos-v1.yaml demo list`
`cargo r -- --additional-demos-file demos/demos-v1.yaml demo describe trino-taxi-data`

**Review by maltesander — APPROVED** [review]
LGTM.

**Comment by sbernauer** [comment]
@maltesander could you please also review the new docs?
I think it's a good idea to document any demo, so i created a sub-folder for them.

File: docs/modules/ROOT/pages/commands/demo.adoc [hunk 4]
```diff
@@ -1,3 +1,129 @@
...
+It is tied to a specific stack of the Stackable data platform, which will provide the required products for the demo.
+
+== Browse available demos
+To list the available demos, run the following command:
+
+[source,console]
+----
+$ stackablectl demo list
+DEMO                                STACKABLE STACK           DESCRIPTION
+trino-taxi-data                     trino-superset-s3         Demo loading 2 years of New York taxi data into S3 bucket, creating Trino table and Superset dashboard
```
Comment (maltesander) [inline docs/modules/ROOT/pages/commands/demo.adoc:13]: Did not test it but i think the demo link will be shown here in the description?

```suggestion
trino-taxi-data                     trino-superset-s3         Demo loading 2 years of New York taxi data into S3 bucket, creating Trino table and Superset dashboard. Documentation: https://docs.stackable.tech/stackablectl/stable/demos/trino-taxi-data.html```

File: docs/modules/ROOT/pages/commands/demo.adoc [hunk 5]
```diff
@@ -1,3 +1,129 @@
...
+----
+
+Detailed information of a demo can be queried with the `describe` command:
+
+[source,console]
+----
+$ stackablectl demo describe trino-taxi-data
+Demo:               trino-taxi-data
+Description:        Demo loading 2 years of New York taxi data into S3 bucket, creating Trino table and Superset dashboard
```
Comment (maltesander) [inline docs/modules/ROOT/pages/commands/demo.adoc:22]:
```suggestion
Description:        Demo loading 2 years of New York taxi data into S3 bucket, creating Trino table and Superset dashboard. Documentation: https://docs.stackable.tech/stackablectl/stable/demos/trino-taxi-data.html
```

File: docs/modules/ROOT/pages/demos/trino-taxi-data.adoc [hunk 6]
```diff
@@ -0,0 +1,311 @@
+= trino-taxi-data
+
+This demo will
+
+* Install the needed Stackable operators
```
Comment (maltesander) [inline docs/modules/ROOT/pages/demos/trino-taxi-data.adoc:5]:
```suggestion
* Install the required Stackable operators
```

File: docs/modules/ROOT/pages/demos/trino-taxi-data.adoc [hunk 7]
```diff
@@ -0,0 +1,311 @@
...
+* Spin up the follow data products
+** Superset: A modern data exploration and visualization platform. This demo uses it to execute SQL queries and build dashboards
+** Trino: A fast distributed SQL query engine for big data analytics that helps you explore your data universe. This demo uses it to enable SQL access to the data
+** MinIO: A S3 compatible object store. This demo uses it as long-term storage.
```
Comment (maltesander) [inline docs/modules/ROOT/pages/demos/trino-taxi-data.adoc:9]: Not sure about this one :D
```suggestion
** MinIO: A S3 compatible object store. This demo uses it as persistent storage.
```

File: docs/modules/ROOT/pages/demos/trino-taxi-data.adoc [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+** Hive metastore: A service that stores metadata related to Apache Hive and other services. This demo uses it as metadata storage for Trino
+* Load testdata into S3. It contains 2 and a half years of New York City taxi trips.
```
Comment (maltesander) [inline docs/modules/ROOT/pages/demos/trino-taxi-data.adoc:11]: I would go for either both text or both numbers
```suggestion
* Load testdata into S3. It contains 2.5 years of New York City taxi trips.
```

File: docs/modules/ROOT/pages/demos/trino-taxi-data.adoc [hunk 9]
```diff
@@ -0,0 +1,311 @@
...
+* Make data accessible via SQL in Trino
+* Create Superset dashboards for visualization
+
+You can see the deployed products as well as their relationship in the following diagram:
+
+image::demo-trino-taxi-data/overview.png[]
+
+== Inspect data in S3
+The S3 provided by MinIO provides long-term storage.
```
Comment (maltesander) [inline docs/modules/ROOT/pages/demos/trino-taxi-data.adoc:20]: If you adapt the persistent above, keep it consistent here. I never really heard of long-term storage.

File: docs/modules/ROOT/pages/demos/trino-taxi-data.adoc [hunk 10]
```diff
@@ -0,0 +1,311 @@
...
+  avg(date_diff('minute', tpep_pickup_datetime, tpep_dropoff_datetime)) as avg_duration_min
+from ny_taxi_data
+group by 1
+order by 1
+----
+
+image::demo-trino-taxi-data/superset_7.png[]
+
+== Summary
+The demo loaded 2 and a half years of taxi trip data from New York City with 68 million records and a total size of 1GB in parquet files.
```
Comment (maltesander) [inline docs/modules/ROOT/pages/demos/trino-taxi-data.adoc:103]:
```suggestion
The demo loaded 2.5 years of taxi trip data from New York City with 68 million records and a total size of 1GB in parquet files.
```

File: docs/modules/ROOT/pages/demos/trino-taxi-data.adoc [hunk 11]
```diff
@@ -0,0 +1,311 @@
...
+
+[source]
+----
+ vendorid |  tpep_pickup_datetime   |  tpep_dropoff_datetime  | duration_min | passenger_count | trip_distance | payment_type | fare_amount | tip_amount | total_amount 
+----------+-------------------------+-------------------------+--------------+-----------------+---------------+--------------+-------------+------------+--------------
+        2 | 2021-08-01 19:47:43.000 | 2021-08-01 19:57:54.000 |           10 |           112.0 |           1.8 | Credit card  |         9.0 |       2.46 |        14.76 
+        2 | 2021-08-03 11:51:58.000 | 2021-08-03 12:09:29.000 |           17 |            96.0 |          1.56 | Credit card  |        11.5 |       2.22 |        17.02 
+----
+Pretty cheap for that amount of persons!
+This probably is a invalid record.
```
Comment (maltesander) [inline docs/modules/ROOT/pages/demos/trino-taxi-data.adoc:207]:
```suggestion
This probably is an invalid record.
```

File: docs/modules/ROOT/pages/quickstart.adoc [hunk 12]
```diff
@@ -1,5 +1,144 @@
...
+Stackable provides a set of ready-to-use demos.
+As of writing (2022/09/10) only a single demo is available, but further demos will be added in the future.
+They will automatically show up, as `stackablectl` fetches the available list of demos via the Internet.
+To list the available demos run the following command:
+
+[source,console]
+----
+$ stackablectl demos list
+DEMO                                STACKABLE STACK           DESCRIPTION
+trino-taxi-data                     trino-superset-s3         Demo loading 2 years of New York taxi data into S3 bucket, creating Trino table and Superset dashboard
```
Comment (maltesander) [inline docs/modules/ROOT/pages/quickstart.adoc:19]: Description change should be reflected here

**Review by maltesander — CHANGES_REQUESTED** [review]
I like it but im not sure about the detail of the description. It feels a lot of it could e.g. be in the superset docs (the story about how many passengers, invalid records, views etc.)

**Comment by sbernauer** [comment]
I moved the documentation to a separate (optional) field so it's machine-readable. I don't think we want to print the docs link when listing the demos, what do you think?
```
$ cargo r -- --additional-demos-file demos/demos-v1.yaml demo list
DEMO                                STACKABLE STACK           DESCRIPTION
trino-taxi-data                     trino-superset-s3         Demo loading 2 years of New York taxi data into S3 bucket, creating Trino table and Superset dashboard
$ cargo r -- --additional-demos-file demos/demos-v1.yaml demo describe trino-taxi-data
Demo:               trino-taxi-data
Description:        Demo loading 2 years of New York taxi data into S3 bucket, creating Trino table and Superset dashboard
Documentation:      https://docs.stackable.tech/stackablectl/stable/demos/trino-taxi-data.html
Stackable stack:    trino-superset-s3
Labels:             trino, superset, minio, s3, ny-taxi-data
```

File: docs/modules/ROOT/pages/demos/trino-taxi-data.adoc (hunk 7 above)
Comment (sbernauer) [reply docs/modules/ROOT/pages/demos/trino-taxi-data.adoc:9]: Changed to `** MinIO: A S3 compatible object store. This demo uses it as a persistent storage to store all the data used`

File: docs/modules/ROOT/pages/demos/trino-taxi-data.adoc (hunk 9 above)
Comment (sbernauer) [reply docs/modules/ROOT/pages/demos/trino-taxi-data.adoc:20]: Changed to `The S3 provided by MinIO is used as persistent storage to store all the data used`

**Comment by sbernauer** [comment]
Thanks @maltesander, comments should be addressed.
Regarding the level of detail: I plan to document all the demos this way as i don't assume any(or not much) knowledge of the users. The `(the story about how many passengers, invalid records, views etc.)` is for me bound to this specific datasets and in turn to this specific demo. I'm planning on adding a similar demo with other data, maybe real-time data with druid. The the analysis  in Superset will be a different story.
Anyway it might be a good idea to mention and link to the demos out there using Superset. Something like "If you'd like to see Superset i action check out this demos: [.,..]". But I'm hoping that demos will be more present in the future and some sort of an entry-point for the users ;)

**Comment by maltesander** [comment]
> Thanks @maltesander, comments should be addressed. Regarding the level of detail: I plan to document all the demos this way as i don't assume any(or not much) knowledge of the users. The `(the story about how many passengers, invalid records, views etc.)` is for me bound to this specific datasets and in turn to this specific demo. I'm planning on adding a similar demo with other data, maybe real-time data with druid. Then the analysis in Superset will be a different story. Anyway it might be a good idea to mention and link to the demos out there using Superset. Something like "If you'd like to see Superset i action check out this demos: [.,..]". But I'm hoping that demos will be more present in the future and some sort of an entry-point for the users ;)

Ok yeah related to the demo data that makes sense then for me. One more thing: 
In the `stackablectl demo describe` examples in the docs the "documentation" field does not show up yet?

**Comment by sbernauer** [comment]
Oh noes, missed that. Added now

**Review by maltesander — APPROVED** [review]
Nice! LGTM.

**Comment by sbernauer** [comment]
Thanks a lot for your very fast review!

**Pull request closed without merging**